File: Task/Queue-Definition/Nim/queue-definition.nim
type

  Node[T] = ref object
    value: T
    next: Node[T]

  Queue*[T] = object
    head, tail: Node[T]
    length: Natural

func initQueue*[T](): Queue[T] = Queue[T]()

func len*(queue: Queue): Natural =
  queue.length

func isEmpty*(queue: Queue): bool {.inline.} =
  queue.len == 0

func push*[T](queue: var Queue[T]; value: T) =
  let node = Node[T](value: value, next: nil)
  if queue.isEmpty: queue.head = node
  else: queue.tail.next = node
  queue.tail = node
  inc queue.length

func pop*[T](queue: var Queue[T]): T =
  if queue.isEmpty:
    raise newException(ValueError, "popping from empty queue.")
  result = queue.head.value
  queue.head = queue.head.next
  dec queue.length
  if queue.isEmpty: queue.tail = nil


when isMainModule:

  var fifo = initQueue[int]()

  fifo.push(26)
  fifo.push(99)
  fifo.push(2)
  echo "Fifo size: ", fifo.len()
  try:
    echo "Popping: ", fifo.pop()
    echo "Popping: ", fifo.pop()
    echo "Popping: ", fifo.pop()
    echo "Popping: ", fifo.pop()
  except ValueError:
    echo "Exception catched: ", getCurrentExceptionMsg()

File: Task/Queue-Definition/PascalABC.NET/queue-definition.pas
type
  Node<T> = auto class
    data: T;
    next: Node<T>;
  end;
  MyQueue<T> = class
    head,tail: Node<T>;
  public
    procedure Enqueue(x: T);
    begin
      if tail = nil then
      begin
        tail := new Node<T>(x,nil);
        head := tail;
      end
      else
      begin
        tail.next := new Node<T>(x,nil);
        tail := tail.next;
      end;
    end;
    function Dequeue(): T;
    begin
      Result := head.data;
      head := head.Next;
      if head = nil then
        tail := nil;
    end;
    function IsEmpty: boolean := head = nil;
  end;

begin
  var q := new MyQueue<integer>;
  for var i:=1 to 5 do
    q.Enqueue(i);
  while not q.IsEmpty do
    Print(q.Dequeue);
  Print(q.IsEmpty);
end.
